### ISSUE TITLE
Keep master stable and use a dev branch?

### ISSUE THREAD

**Issue by vanhauser-thc** [AFLplusplus/AFLplusplus#245]
In the last week we did a major change to afl++ that has a high impact and can have yet unknown consequences. And it broke our master for over a day.

We develop a lot in the master, so sometimes the master is broken for some time.
For people who clone, compile and work from master this can be problematic as they might get a state that has a big bug or doesnt compile.
So a solution for that would be that the master is basically the same as the current release and we put all changes to a/"the" dev branch, and for a new release we pull the changes to the master

On the other hand people might want the bleeding edge because we really put out great enhancements all the time.
And cloning the git in other repos (thinking of dyninst or dynamorio) can fail too because of changes they do, so that might be okay for people and for those who want to be sure everything is fine would use the .tar.gz of a release.

So ... what should we do?
This is rather a question for our users than ourselves :) On the other hand I doubt many people will read and comment on this issue ...

Well, lets discuss this.

**Comment by andreafioraldi** [comment]
As I said before, a devel branch is the right choice to take. People use AFL++ mainly cloning master, we got issues about dead bugs fixed in master by people that cloned master just one day before the fix many times.

**Comment by domenukk** [comment]
I second this.
- only push to dev and accept PRs against dev
- only merge/rebase into master if Travis is happy with dev
- lock down master in github settings

**Comment by domenukk** [comment]
(or alternatively have a `stable` branch and communicate this fact in the docs)

**Comment by hexcoder-** [comment]
Ok, one question would be, when is new code considered to be stable? Without rigerous testing that is sometimes difficult  to decide.
So we would have several levels then:
 
- The releases (which should be very stable)

- The master branch, which consolidates the dev branch and passes `make tests` on travis

- The development branch for active work in progress

- Other experimental branches for concepts which might not work out

Thats fine with me.

**Comment by h1994st** [comment]
I agree with the design of multi-level branches.

Here is a very good example and commonly used by many software teams: [Gitflow Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow). A quick summary:

> The overall flow of Gitflow is:
>
> 1. A `develop` branch is created from `master`
> 2. A `release` branch is created from `develop`
> 3. Feature branches are created from `develop`
> 4. When a feature is complete it is merged into the `develop` branch
> 5. When the release branch is done it is merged into `develop` and master
> 6. If an issue in `master` is detected a `hotfix` branch is created from master
> 7. Once the `hotfix` is complete it is merged to both `develop` and `master`

We can take a look at it. This workflow model is also shipped with a tool called `git-flow`. If necessary, we can transition to it.

**Comment by domenukk** [comment]
I'm pro `dev` instead of `develop`, less typing/tabbing

**Comment by h1994st** [comment]
> I'm pro `dev` instead of `develop`, less typing/tabbing

Yeah, I agree.

**Comment by domenukk** [comment]
So all in favor. @vanhauser-thc, make the changes if you find the time :)

**Comment by vanhauser-thc** [comment]
I dont have the time, I am leaving tomorrow for a vacation without laptop - you guys make the necessary changes :) you are all grown up! :p

**Comment by domenukk** [comment]
Enjoy

**Comment by vanhauser-thc** [comment]
Will do so mid of the week

**Comment by vanhauser-thc** [comment]
So this is going down as follows: 

step 1: creating a new organization - DONE
step 2: Dominik (DONE), @hexcoder- and @andreafioraldi join the organization
step 3.1: I move the repository to the organization and
step 3.2: I create the branch "dev"
step 3.3: I update all pointers in the repo to the new repo location
step 3.4: I update the readme to describe the dev/trunk concept

steps 3.1-3.4 will have to happen on a very short time frame
from this point on only ever commit to dev or a experimental branch.

I will see if I can somehow lock down the trunk/master branch so nobody . Anybody knows this maybe?

**Comment by domenukk** [comment]
Github allows you to lock down master in the repo settings.
Also I'd stick with calling it master, trunk is _very_ svn 🙃

**Comment by vanhauser-thc** [comment]
> Also I'd stick with calling it master, trunk is _very_ svn 🙃
guess what I am using :p

**Comment by vanhauser-thc** [comment]
moving, renaming, etc. all done.

I did not create a release branch because we have release downloads and it would mean managing one more branch ...

**Comment by vanhauser-thc** [comment]
I modified the setup so a review is required to pull from dev to master - lets see if it works.

any other things that still have to be done?
btw I gave you all admin status.

@andreafioraldi @domenukk @hexcoder-

**Comment by domenukk** [comment]
The main branch, if cloning from aflplusplus/aflplusplus is now dev.
I think this should still be the master branch, else a simple clone && make
will still fail.

Apart from that, it looks good :)

**Comment by vanhauser-thc** [comment]
@domenukk ah damn. I dont know how to config this then so that PRs always go to dev but master is still the main.
you have admin perms and maybe can set it up like it should be.

alternative would be that every PR has to be reset by hand to dev ...

**Comment by vanhauser-thc** [comment]
I think we can close this now?
If not just reopen or put comments here :)
